### PR TITLE
:art: Refractor `storage_helpers` module

### PR DESCRIPTION
**Pull request by acemasterjb**
## Summary
Applied refactor of `storage_helpers` according to [recommendations from an earlier PR](https://github.com/butterdao/whip/pull/88#discussion_r932190774).

## Details
The `storage_helpers` module is now more appropriately divided and placed into distinctive modules according to what function it serves. Including `whitelist` and `treasury_cacher` logic.

### CODE REVIEW

File: backend/app/token_whitelists/whitelists.py [hunk 1]
```diff
@@ -7,42 +7,54 @@
...
+    retrieve_token_whitelist,
+    store_token_whitelist,
+)
 
 COVALENT_KEY = os.getenv("COVALENT_KEY")
 COVALENT_POOLS_URL = "https://api.covalenthq.com/v1/{chain_id}/xy=k/{protocol}/pools/"
 
 
-async def _get_covalent_pair_list(
+async def get_single_covalent_pairs_page(
```
Comment (lajarre) [inline backend/app/token_whitelists/whitelists.py:19]: A good naming for a generator function is a description of what it produces (yields). I'd add that using plural is good for generators.

Here, `single` and `page` are both confusing and don't describe the output. Let's remove them.

File: backend/app/token_whitelists/__init__.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+from .whitelists import (
```
Comment (lajarre) [inline backend/app/token_whitelists/__init__.py:1]: All imports in this file are really exports from the `token_whitelists` application, that are useful for other external apps communicating with it. It seems `get_covalent_pair_list` for example doesn't need being exported, etc.

Exporting a function basically means that other apps can start depending on this given function. We should thus strive to export the minimum.

(I'm using the "application" term as a reference to hexagonal apps: each of `treasury`, `spread` and `token_whitelists` are applications, or maybe "app-modules")

File: backend/app/token_whitelists/storage_helpers/storage_helpers.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+from typing import Union
```
Comment (lajarre) [inline backend/app/token_whitelists/storage_helpers/storage_helpers.py:1]: We started using the `adapter` subfolder in other app-modules (`treasury`, `spread`), to list modules that take care of adapters. These adapter modules are named by the external service they are relating to.

Here, `storage_helpers.py` is really an adapter to Redis. I thus suggest: `token_whitelists/adapters/redis.py`. No need for a folder, btw.

File: backend/app/treasury/storage_helpers/storage_helpers.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+import json
```
Comment (lajarre) [inline backend/app/treasury/storage_helpers/storage_helpers.py:1]: Same remark as in backend/app/token_whitelists/storage_helpers/storage_helpers.py

File: backend/app/treasury/tests/test_treasury_cacher.py [hunk 5]
```diff
@@ -4,11 +4,11 @@
 from httpx import Client
 from pytest import MonkeyPatch, mark, raises
 
-from backend.app.libs.storage_helpers.storage_helpers import (
+from backend.app.treasury.storage_helpers.storage_helpers import (
```
Comment (lajarre) [inline backend/app/treasury/tests/test_treasury_cacher.py:7]: Cool thing now that you're in an independent app-module (`treasury`), neat relative imports are now allowed: `from ..storage_helpers.storage_helpers import …`

File: backend/app/treasury/storage_helpers/__init__.py [hunk 6]
```diff
@@ -0,0 +1,2 @@
+from .storage_helpers import retrieve_treasuries_metadata
+from .treasury_cacher import get_and_store_treasury_list
```
Comment (lajarre) [inline backend/app/treasury/storage_helpers/__init__.py:2]: (For some reason I'm not able to comment on the `treasury_cacher.py` file so here is my remark on this file)

I have the same remark as in backend/app/token_whitelists/storage_helpers/storage_helpers.py  
The treasury_cacher.py module is really an adapter to cryptostats, so it should be named accordingly.

Another important remark is that this adapter should take care only about cryptostats, not anything else. So we should not allow it to make calls to the `store_treasuries_metadata` function which stems from another adapter. As a rule of thumb, whatever function is trying to "merge" several adapter together should be in the core of the application. More concretely, `get_and_store_treasury_list` should be in your core module, which is `whitelists.py`.

**Review by lajarre — CHANGES_REQUESTED** [review]

File: backend/app/treasury/storage_helpers/__init__.py (hunk 6 above)
Comment (lajarre) [reply backend/app/treasury/storage_helpers/__init__.py:2]: I'm realizing that we're not respecting this rule in the rest of the codebase, as we're accessing the "global" `db` object. That basically means that our other adapters are already too knowledgeable. I'm not sure yet how to address the problem of the global `db` object.

Comment (lajarre) [reply backend/app/treasury/storage_helpers/__init__.py:2]: > Do you see this as issue due to the fact that we are importing the db into the modules like from .... import db or that we're using a "global db" in the first place?

My concern is with the code: `import db`, not really with the existence of a global object.

We can totally have a singleton db object if that helps.

File: backend/app/treasury/tests/test_make_transfers_for_treasury.py [hunk 7]
```diff
@@ -5,13 +5,14 @@
...
 import pytest
+import pytest_asyncio
 from pytz import UTC
 
 from .. import actions
 from ..models import ERC20, Transfer, Treasury
 
 
-@pytest.fixture
+@pytest_asyncio.fixture
```
Comment (lajarre) [inline backend/app/treasury/tests/test_make_transfers_for_treasury.py:15]: :+1:

**Review by lajarre — APPROVED** [review]
Awesome! We're getting a much tidier codebase.

Next challenges I see:

- `treasury` should not export its own adapters. One approach could be that Celery tasks are defined in `treasury/tasks.py` and somehow registered inside the global `celery_app`.
- as "adapters" are getting more complex, we might need to look into splitting down (probably adapters vs ports) more intelligently so as to play better with our caching logic.

To be refined later.

File: backend/app/treasury/tests/test_make_transfers_for_treasury.py (hunk 7 above)
Comment (acemasterjb) [reply backend/app/treasury/tests/test_make_transfers_for_treasury.py:15]: > Next challenges I see:
> 
> * `treasury` should not export its own adapters. One approach could be that Celery tasks are defined in `treasury/tasks.py` and somehow registered inside the global `celery_app`.
> * as "adapters" are getting more complex, we might need to look into splitting down (probably adapters vs ports) more intelligently so as to play better with our caching logic.

Interesting, I'll definitely like to continue with this conversation.

**Pull request merged**